### ISSUE TITLE
[Stop criteria] How does 【parameter optimizer】 work in Chaldene AI and in NN evaluation?

### ISSUE THREAD

**Issue by xp1632** [xp1632/Aalen_working_log#61]
### 1. Convergence Threshold (Primary)
`running_loss_delta < 1e-5  # 0.00001`
- Tracks the average change in loss over the last 5 epochs
- Stops when Loss changes become smaller than 0.00001

---

### 2. Maximum Epochs (Safety Limit)
`epoch >= max_epochs  # Default: 100`

- Hard limit to prevent infinite training

---

**3. No Learnable Parameters (Edge Case)**
```
if len(list(model.parameters())) == 0:
    return  # Skip training
```
- If workflow has no trainable parameters
- Example: Only fixed operations like RgbToGrayscale, Invert

---

### 4. Non-Differentiable Operations (Gradient Blockage)
```
if not output.requires_grad:
    return  # Can't backpropagate
```
- Detected during forward pass
- If any operation blocks gradients
- Example: Using AdjustBrightness instead of DiffBrightness

**Comment by xp1632** [comment]
### Information in ori repo

<img width="1481" height="482" alt="Image" src="https://github.com/user-attachments/assets/fb3f96be-0046-41a1-9b93-2703e1d611ea" />

**Comment by xp1632** [comment]
<html>
<body>
<html><head></head><body><h1>Convergence &amp; Early Stopping: NN Baseline vs LLM+Gradient</h1>
<h2>Comparison Table</h2>

Aspect | LLM+Gradient (Ours) | NN Baseline (CNN / UNet)
-- | -- | --
Parameters | 2–5 (semantic: brightness, gamma, etc.) | ~11K (CNN) / ~1M (UNet)
Optimization target | Kornia operation parameters | All network weights
Convergence criterion | Loss delta < 1e-5 over last 5 epochs | Smoothed loss (10-epoch window) no improvement for 50 epochs
Max epochs (safety limit) | 100 | 1000
Typical convergence | 10–30 epochs | 100–500 epochs
Best model selection | Final (convergence is smooth) | Best checkpoint restored
Additional stopping conditions | No learnable params → skip; Non-differentiable ops → skip | —
Optimizer | Adam, lr=0.1 | Adam, lr=0.001
Loss function | MSE (input vs target) | MSE (input vs target)


<h2>Why Different Stopping Criteria Are Valid</h2>
<p>The LLM+Gradient method and neural network baselines are fundamentally different optimization problems, and applying the same stopping rule to both would be less fair than using method-appropriate criteria.</p>
<p><strong>LLM+Gradient</strong> optimizes 2–5 semantic parameters (e.g., brightness factor, gamma value) via gradient descent on a smooth, low-dimensional loss landscape. Convergence is fast and stable — a tight threshold (loss delta &lt; 1e-5 over 5 epochs) reliably detects when the optimum is reached, typically within 10–30 epochs. The max epoch limit of 100 is rarely reached.</p>
<p><strong>Neural network baselines</strong> optimize hundreds of thousands to millions of weights using mini-batch gradient descent. The loss signal is inherently noisier due to mini-batching, so a tight delta threshold would trigger premature stopping. Instead, a patience-based approach (50 epochs without improvement on a 10-epoch smoothed loss) is standard practice and generous to the baseline. The best model checkpoint is restored after stopping, ensuring the reported result reflects peak performance rather than a potentially degraded final state.</p>
<p>The key principle is that <strong>each method should be given every reasonable opportunity to reach its best achievable performance</strong>. The NN baseline receives 10× more maximum epochs, 10× more patience, and best-checkpoint selection — all of which are generous advantages. Despite this, the core finding remains: a method-appropriate stopping criterion for each approach ensures a fair comparison where performance differences reflect the fundamental strengths of the approach, not artifacts of training procedure.</p>
<h2>Suggested Paper Text</h2>
<blockquote>
<p>All neural network baselines were trained with Adam (lr=0.001) for up to 1,000 epochs with early stopping (patience=50 epochs, smoothed over a 10-epoch window), and the best checkpoint was restored for evaluation. Our LLM+Gradient method optimizes workflow parameters with a convergence threshold of 1e-5 on the running loss delta, typically converging within 10–30 epochs. Both methods minimize MSE loss between predicted and target images. The differing stopping criteria reflect the distinct optimization landscapes: our method optimizes 2–5 semantic parameters on a smooth loss surface, while neural networks optimize ~10^5–10^6 weights with noisy mini-batch gradients, requiring greater patience to reach their optimal performance.</p>
</blockquote></body></html>
</body>
</html>

**Comment by xp1632** [comment]
<img width="1151" height="301" alt="Image" src="https://github.com/user-attachments/assets/e32693ea-a7c0-479d-92a7-8452b9c4487f" />
